### PR TITLE
Better file conflict message

### PR DESCRIPTION
**Pull request by sergiusens**
When parts conflict the message is plain and blunt, we
should do some hand holding.

LP: #1617390

Signed-off-by: Sergio Schvezov <sergio.schvezov@ubuntu.com>

### CODE REVIEW

File: snapcraft/internal/errors.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+class SnapcraftError(Exception):
+    """Base class for all storeapi exceptions.
```
Comment (evandandrea) [inline snapcraft/internal/errors.py]: Copy and paste typo from `storeapi/errors.py`?

File: snapcraft/internal/errors.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+class MissingState(Exception):
+    pass
+
+
+class SnapcraftPartConflictError(SnapcraftError):
+
+    fmt = (
+        'Parts {other_part_name!r} and {part_name!r} have the following file '
+        'paths in common which have different contents:\n'
+        '{contents}\n\n'
```
Comment (evandandrea) [inline snapcraft/internal/errors.py]: I found it confusing to read about file contents and then see a variable called contents that was really a set of paths.

**Comment by sergiusens** [comment]
thanks @evandandrea I'll apply some fixes from your comments. Is the messaging ok now or think it still needs improvements?

File: snapcraft/internal/errors.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/internal/errors.py]: indeed

**Comment by evandandrea** [comment]
The messaging is much improved, thank you.

File: snapcraft/internal/errors.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        'paths in common which have different contents:\n'
+        '{file_paths}\n\n'
+        'Snapcraft offers some capabilities to solve this the following '
+        'keywords:\n'
+        '- `filesets`\n'
+        '- `stage`\n'
+        '- `snap`\n'
+        '- `organize`\n\n'
+        'Learn more about these part keywords by running '
+        '`snapcraft help plugins`'
```
Comment (kyrofa) [inline snapcraft/internal/errors.py]: This message is wonderful, but if there are more than a few files in that conflict list I could see it becoming less digestible. A little whitespace might help-- compare this:

```
Parts {other_part_name!r} and {part_name!r} have the following file 'paths in common which have different contents:
/home/foo/bar/baz/parts/qux/install/usr/include/my_include1.h
/home/foo/bar/baz/parts/qux/install/usr/include/my_include2.h
/home/foo/bar/baz/parts/qux/install/usr/include/my_include3.h
/home/foo/bar/baz/parts/qux/install/usr/include/my_include4.h

Snapcraft offers some capabilities to solve this the following keywords:
- `filesets`
- `stage`
- `snap`
- `organize`

Learn more about these part keywords by running `snapcraft help plugins`
```

With this:

```
Parts {other_part_name!r} and {part_name!r} have the following file 'paths in common which have different contents:
    /home/foo/bar/baz/parts/qux/install/usr/include/my_include1.h
    /home/foo/bar/baz/parts/qux/install/usr/include/my_include2.h
    /home/foo/bar/baz/parts/qux/install/usr/include/my_include3.h
    /home/foo/bar/baz/parts/qux/install/usr/include/my_include4.h

Snapcraft offers some capabilities to solve this the following keywords:
    - `filesets`
    - `stage`
    - `snap`
    - `organize`

Learn more about these part keywords by running `snapcraft help plugins`
```

**Comment by sergiusens** [comment]
@kyrofa point taken!

**Comment by kyrofa** [comment]
:+1: from me!

File: integration_tests/test_stage.py [hunk 4]
```diff
@@ -30,7 +30,19 @@ def test_conflicts(self):
...
+            'Snapcraft offers some capabilities to solve this by use '
+            'of the following keywords:\n'
+            '    - `filesets`\n'
+            '    - `stage`\n'
+            '    - `snap`\n'
+            '    - `organize`\n\n'
+            'Learn more about these part keywords by running '
+            '`snapcraft help plugins`'
+        )
+        self.assertThat(exception.output, Contains(expected_help))
```
Comment (elopio) [inline integration_tests/test_stage.py]: update the copyright year, please

File: snapcraft/internal/errors.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        '    - `filesets`\n'
+        '    - `stage`\n'
+        '    - `snap`\n'
+        '    - `organize`\n\n'
+        'Learn more about these part keywords by running '
+        '`snapcraft help plugins`'
+    )
+
+    def __init__(self, *, part_name, other_part_name, conflict_files):
+        spaced_conflict_files = ('    {}'.format(i) for i in conflict_files)
```
Comment (elopio) [inline snapcraft/internal/errors.py]: This feels like should be handled by the format, but I don't know how to do it. What about overwriting __str__ and making the transformation from list to list with spaces in there?

File: snapcraft/storeapi/errors.py [hunk 6]
```diff
@@ -14,25 +14,9 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-
 from simplejson.scanner import JSONDecodeError
 
-
-# TODO move to snapcraft.errors --elopio - 2016-06-20
```
Comment (elopio) [inline snapcraft/storeapi/errors.py]: :dancers:

File: snapcraft/internal/errors.py (hunk 5 above)
Comment (sergiusens) [inline snapcraft/internal/errors.py]: I don't think that would add clarity to what is going on. Maybe a different PR after experimenting.

**Pull request merged**